compiler/passes/ResolveScope.cpp

```diff
@@ -529,72 +528,147 @@ Symbol* ResolveScope::followImportUseChains(UnresolvedSymExpr* expr) const {
   }
 }
 
+static const char* getNameFrom(Expr* e) {
+  const char* name = NULL;
+  if (UnresolvedSymExpr* uSym = toUnresolvedSymExpr(e)) {
+    return uSym->unresolved;
+  } else if (SymExpr* se = toSymExpr(e)) {
+    if (get_string(se, &name) == true)
+      return name;
+  }
+
+  return NULL;
+}
+
+
+// Finds the first name (other than this/super) in the Expr
+//   name is that name
+//   call is the call contaning the name on the left branch or NULL
+//     (which is the call requiring further processing)
+//   scope is the scope indicated by any this. / super. or NULL
+void ResolveScope::firstImportedModuleName(Expr* expr,
+                                           const char*& name,
+                                           CallExpr*& call,
+                                           const ResolveScope*& scope) const {
+  if (const char* n = getNameFrom(expr)) {
+    if (n == astrThis) {
+      scope = this;
+    } else if (n == astrSuper) {
+      if (mParent == NULL || mParent->mParent == NULL)
+        USR_FATAL(expr, "cannot import super from a toplevel module");
+      scope = mParent;
+    } else {
+      name = n;
+    }
+  } else if (CallExpr* c = toCallExpr(expr)) {
+    if (c->isNamedAstr(astrSdot) == true) {
+      const char* oldName = name;
+      const ResolveScope* curScope = this;
+      if (scope != NULL) curScope = scope;
+      curScope->firstImportedModuleName(c->get(1), name, call, scope);
+      if (call == NULL && name != NULL && oldName == NULL) {
+        // update the call the first time the left-branch set the name.
+        call = c;
+      }
+      if (name == NULL) {
+        // look in the right branch in case of e.g. super.super.M
+        // to update scope
+        if (scope != NULL) curScope = scope;
+        curScope->firstImportedModuleName(c->get(2), name, call, scope);
+      }
+
+    } else {
+      INT_FATAL("expected a . call");
+    }
+  } else {
+    INT_FATAL("expected a name or a call");
+  }
+}
+
 Symbol* ResolveScope::lookupForImport(Expr* expr) const {
   Symbol* retval = NULL;
 
-  if (UnresolvedSymExpr* uSym = toUnresolvedSymExpr(expr)) {
+  const char* name = NULL;
+  CallExpr* call = NULL;
+  const ResolveScope* relativeScope = NULL;
+
+  firstImportedModuleName(expr, name, call, relativeScope);
+  if (name == NULL)
+    USR_FATAL(expr, "'this.' can only be used as prefix of import");
```

Comment:
I feel like this isn't the only situation that could make name remain NULL, but I'm not thinking of other situations at the moment.  I'll take a look at the tests and see what else I think should be covered